Class name highlighting

Wilfred opened the pull request:
Highlight class names in Typescript files, and add a test.

Before:

![screen shot 2018-05-09 at 11 12 37](https://user-images.githubusercontent.com/70800/39809569-0c78215c-537a-11e8-901f-722abbc4d884.png)

After:

![screen shot 2018-05-09 at 11 12 07](https://user-images.githubusercontent.com/70800/39809575-14d8d030-537a-11e8-954c-4a314de054fd.png)

I've also removed a few bits of unused code. These are separate commits, so you can see what's going on :)

Wilfred commented on typescript-mode.el:
```diff
@@ -1985,9 +1985,7 @@ This performs fontification according to `typescript--class-styles'."
-  '(typescript--font-lock-keywords-3 typescript--font-lock-keywords-1
-                                   typescript--font-lock-keywords-2
-                                   typescript--font-lock-keywords-3)
+  '(typescript--font-lock-keywords-3)
```
Woah, I had no idea you could do that. OK, I've removed this change :)

Wilfred commented on typescript-mode.el:
```diff
@@ -626,11 +626,6 @@ seldom use, either globally or on a per-buffer basis."
-  "*Hook called by `typescript-mode'."
-  :type 'hook
-  :group 'typescript)
-
```
Reverted.

Wilfred commented on typescript-mode.el:
```diff
@@ -1772,6 +1773,12 @@ and searches for the next token to be highlighted."
+
+    (,(rx "implements" (+ space) (group (+ (or (syntax word) (syntax symbol)))))
+     (1 font-lock-type-face))
+
```
This ensures that `class Foo implements Bar` leads to `Bar` being highlighted as a type. The unit test I wrote covers this case. Whilst it's not perfect (e.g. `implements /* foo */ Bar` won't be highlighted correctly), it covers the common cases.

Are you happy with this, or would you like me to change it in some way?

lddubeau replied:
What needed changing was in the comment I made summing up the review and reflecting on what highlighting should do. You've made the changes so there's nothing more needed.

josteink commented:
First off: Thanks to @Wilfred for the PR!

As usual, @lddubeau has some good, extensive and well thought through commentary.

I'll try to weigh in my opinions too, but I'm (as always) short on time, so I won't be able to be as thorough. Sorry about that.

> There's a broader issue that the highlighting changes currently proposed are internally inconsistent. 

I agree consistency is an important goal, but I may not be *as* demanding as lddubeau, or maybe I'll just frame it as being more pragmatical: Ensuring we support 100% of TypeScript syntax 100% precisely is a hard job, because TypeScript syntax is flexible and difficult to parse correctly, at least based on regexps.  

> Conversely, in the same string, Glapotis is not a class but an interface and yet it is highlighted as a class name

I think the convention in lots of IDEs and editors is that these represent *type-names*, not specifically class-names or interface-names. As such, I'm fine with both interface-names and class-names being treated similarly.

> So far, typescript-mode.el's fontification has been able (with few a exceptions) to be consistent by aiming for a relatively superficial level of analysis. It identifies syntactic structures and does not go much further beyond this.

Indeed. But I don't think stems from some grand plans or designs, but is just us never having taken the time to refine the implementation beyond the bare minimum.

I don't mind raising the ambitions beyond that, if anyone is willing to put up the code. As such, great initiative @Wilfred :)

> If the mode is going to determine that some identifier should be fontified as class name, it should be fontifying the same identifier as a class name everywhere it is used. In the example above, I'd want Moo to be fontified as a class name in function q(a: Moo): void.

This probably requires more effort than you've put in so far, but I do agree that at least this level of consistency would be nice to get right. Do you think you would be able to add that?

While beyond these cases, there may be more scenarios where we encounter class-names which are (yet) to be highlighted consistently. I'm OK with leaving this until later... When we encounter them.

>  If a class name is fontified as a class name only sometimes, then I cannot generally rely on fontification to know whether what I'm looking at is a class name, and fontifying class names differently from other identifiers does not help me understand the code I'm reading.

While kinda I agree with this, I know there are other major-modes which suffers from inconsistent highlighting, in the sense that classes and types and other things are not properly identified within complex grammar.

That does IMO not take away the usefulness of the major-mode identifying key components of the code *most* of the places you encounter it.

Regarding https://github.com/ananthakumaran/typescript.el/commit/a089a83f6ad50bb1530703cccc9339a5b4e95c9c...

I think that may have been about trying to reduce the amount of byte-compilation warnings? Could you verify if we get more or the same amount of warnings when doing `make` with or without these changes?

Wilfred commented:
Thanks for the thorough feedback! :)

**Byte-compilation warnings**: I removed references to libraries that are entirely unused. `typescript-mode.el` has three warnings (about `font-lock-fontify-buffer`, `syntax-begin-function`, `flyspell-generic-progmode-verify`) both before and after a089a83.

**Highlighting precision**: I believe `js2-mode` is the only major-mode with a full parser implementation. I think these changes are a concrete improvement, and consistent with other major modes. 100% perfect parsing is rare in practice.

I agree a class name is different from an interface, but I think using `font-lock-type-face` is correct, as it's consistent with other Emacs modes (e.g. `java-mode`).

I've added highlighting for type names in definitions with `extends`, `interface` and `enum` too.  I'm reluctant to add highlighting of types of arguments in this PR as it's big enough already, I believe it'll be a little more work, and I'd like to start using this new highlighting soon :)

lddubeau commented:
I've squashed the last two commits together since the 2nd one was really fixing issues with the 1st and manually pushed to `master`. So this is effectively "merged". Thanks again.

As to the broader issue of what highlighting should do, I think we'll just have to "agree to disagree", as people say. I'm pretty sure my opinion is in the minority.
